src/DotLiquid.Tests/Util/NumericConverterTests.cs

```diff
@@ -0,0 +1,189 @@
+        }
+
+        [Test]
+        [TestCaseSource(nameof(ErrorTestCaseSource))]
+        public void TestTryCoerceToNumericTypeErrors(object input, IFormatProvider formatProvider)
+        {
+            bool converted = NumericConverter.TryCoerceToNumericType(input, formatProvider, out object convertedValue);
+            Assert.That(converted, Is.False, $"convertedValue: {convertedValue}");
+        }
+
+        [Test]
+        [TestCaseSource(nameof(GoodTestCaseSource))]
+        public void TestTryParseToNumericType(string input, IFormatProvider formatProvider, object expectedValue)
+        {
+            bool converted = NumericConverter.TryParseToNumericType(input, formatProvider, out object convertedValue);
+            Assert.That(converted, Is.True);
+            Assert.That(convertedValue, Is.EqualTo(expectedValue));
+            Assert.That(convertedValue, Is.TypeOf(expectedValue.GetType()));
+        }
+
+        [Test]
+        [TestCaseSource(nameof(ErrorTestCaseSource))]
+        public void TestTryParseToNumericTypeErrors(string input, IFormatProvider formatProvider)
+        {
+            bool converted = NumericConverter.TryParseToNumericType(input, formatProvider, out object convertedValue);
+            Assert.That(converted, Is.False, $"convertedValue: {convertedValue}");
+        }
+
+        [Test]
+        [TestCaseSource(typeof(NumericHelper), nameof(NumericHelper.GetNumericTypeCombinations))]
+        public void TestNumericCombinationsResultInUpgrade(ValueTuple<Type, Type> types)
+        {
+            var t1 = types.Item1;
+            var t2 = types.Item2;
+            var result = NumericConverter.GetBinaryResultType(t1, t2);
+            Assert.That(result, Is.Not.Null);
+            Assert.That(NumericConverter.GetBinaryResultType(t2, t1), Is.EqualTo(result));
+            Assert.That(TypeLimits[result].Item1 >= TypeLimits[t1].Item1, Is.True);
+            Assert.That(TypeLimits[result].Item1 >= TypeLimits[t2].Item1, Is.True);
+            Assert.That(TypeLimits[result].Item2 <= TypeLimits[t1].Item2, Is.True);
+            Assert.That(TypeLimits[result].Item2 <= TypeLimits[t1].Item2, Is.True);
+        }
+
+        private static IEnumerable GoodTestCaseSource()
+        {
+            IFormatProvider invariantFormatProvider = CultureInfo.InvariantCulture;
+            IFormatProvider frenchFormatProvider = new CultureInfo("fr-FR");
+
+            // Int32
+            yield return new object[] { "0", null, 0 };
+            yield return new object[] { "0", invariantFormatProvider, 0 };
+            yield return new object[] { "123", invariantFormatProvider, 123 };
+            yield return new object[] { "-123", invariantFormatProvider, -123 };
+
+            // Int32 with thousands separator
+            yield return new object[] { String.Format(invariantFormatProvider, "{0:#,##0}", 12567),
+                invariantFormatProvider, 12567 };

```

Comment:
nit: @daviburg would say either place each parameter on its own line or one line. commas sometimes in middle do not make for easy readability.